core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/BaseAsyncAws4Signer.java

```diff
@@ -0,0 +1,106 @@
+import software.amazon.awssdk.auth.signer.params.Aws4SignerParams;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.signer.AsyncRequestBodySigner;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.async.SdkHttpContentPublisher;
+import software.amazon.awssdk.utils.Logger;
+
+
+@SdkInternalApi
+public abstract class BaseAsyncAws4Signer extends BaseAws4Signer implements AsyncRequestBodySigner {
+    private static final Logger LOG = Logger.loggerFor(BaseAsyncAws4Signer.class);
+
+    private static final Pattern AUTHENTICATION_HEADER_PATTERN = Pattern.compile(
+        SignerConstant.AWS4_SIGNING_ALGORITHM + "\\s" + "Credential=(\\S+)" + "\\s" + "SignedHeaders=(\\S+)" + "\\s"
+        + "Signature=(\\S+)");
+
+    @FunctionalInterface
+    protected interface RequestProviderTransformer {
+        SdkHttpContentPublisher transform(SdkHttpContentPublisher publisher);
+    }
+
+    protected BaseAsyncAws4Signer() {
+    }
+
+    @Override
+    public SdkHttpContentPublisher signAsyncRequestBody(SdkHttpFullRequest request,
+                                                        SdkHttpContentPublisher requestPublisher,
+                                                        ExecutionAttributes executionAttributes) {
+        Aws4SignerParams signingParams = extractSignerParams(Aws4SignerParams.builder(), executionAttributes)
+            .build();
+        Aws4SignerRequestParams requestParams = new Aws4SignerRequestParams(signingParams);
+
+        return signAsync(request, requestPublisher, requestParams, signingParams);
+    }
+
+    /**
+     * This method is only used in test, where clockOverride is passed in signingParams
+     */
+    @SdkTestInternalApi
+    protected final SdkHttpContentPublisher signAsync(SdkHttpFullRequest request, SdkHttpContentPublisher requestPublisher,
+                                                Aws4SignerRequestParams requestParams, Aws4SignerParams signingParams) {
+        AwsCredentials sanitizedCredentials = sanitizeCredentials(signingParams.awsCredentials());
+        final byte[] signingKey = deriveSigningKey(sanitizedCredentials, requestParams);
+
+        String headerSignature = getHeaderSignature(request);
+        RequestProviderTransformer transformer = createRequestProviderTransformer(headerSignature, signingKey,
+                                                                                  requestParams, signingParams);
+
+        return transformer.transform(requestPublisher);
+    }
+
+    /**
+     * Create a transformer for requestProvider. The transformer should add signing operator to the original requestProvider
+     *
+     * Can be overriden by subclasses to provide specific signing method
+     */
+    protected abstract RequestProviderTransformer createRequestProviderTransformer(String headerSignature,
```

Comment:
Why not just have this method be the transformer instead of it returning a functional interface?

Reply:
That should do it. I can make the change